System.Memory.Data

ericstj opened the pull request:
Bring in System.Memory.Data to dotnet/runtime.

This ports System.Memory.Data from https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/System.Memory.Data to dotnet/runtime.

I had to react to a couple breaking changes, made formatting changes, and switched to resx.

First change is a patch from azure SDK, just review the second change.

cc @JoshLove-msft

JoshLove-msft committed:
Add BinaryData

Ported from https://github.com/Azure/azure-sdk-for-net/commit/995a13b14c5f635b293c117bfb657d9532b2a202

ericstj committed:
Make System.Memory.Data build and fix tests

ericstj commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -48,10 +48,13 @@ public BinaryData(byte[] data)
         {
+            if (type == null)
+            {
+                type = jsonSerializable == null ? typeof(object) : jsonSerializable.GetType();
```
This change was needed for two reasons.
1. passing null, null now throws in 5.0, so instead pass typeof(object).  https://docs.microsoft.com/en-us/dotnet/core/compatibility/serialization#jsonserializerserialize-throws-argumentnullexception-when-type-parameter-is-null
2. passing object, null never worked and this API is documented as calling `object.GetType()` so I did that.

JoshLove-msft replied:
woops, not sure what happened there - I had this null checking in the prior version. Yes I think we will want to do a patch release for this.

ericstj replied:
Ported here: https://github.com/Azure/azure-sdk-for-net/pull/16957

ericstj commented on src/libraries/System.Memory.Data/test/BinaryDataTests.cs:
```diff
@@ -284,23 +286,24 @@ public async Task StartPositionOfStreamRespectedBackingBuffer(int bufferOffset,
         public void CanCreateBinaryDataFromCustomType()
         {
             TestModel payload = new TestModel { A = "value", B = 5, C = true, D = null };
+            JsonSerializerOptions options = new JsonSerializerOptions { DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull };
```
This was needed because IgnoreNullValues was obsoleted.

ericstj commented on src/libraries/System.Memory.Data/tests/samples/BinaryDataSamples.cs:
```diff
@@ -1,89 +0,0 @@
-using System.Text;
-using Xunit;
-
-namespace Azure.Core.Experimental.Tests.samples
```
These all overlapped over existing tests.

ericstj committed:
Simplify type default

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+        /// <param name="data">The array to wrap.</param>
+        public BinaryData(byte[] data)
+        {
+            if (data == null)
+            {
+                throw new ArgumentNullException(nameof(data));
+            }
+
+            _bytes = data;
```
Nit: this could be:
```suggestion
            _bytes = data ?? throw new ArgumentNullException(nameof(data));
```

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, false).GetAwaiter().GetResult();
```
```suggestion
            return FromStreamAsync(stream, async: false).GetAwaiter().GetResult();
```

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, true, cancellationToken);
```
```suggestion
            return FromStreamAsync(stream, async: true, cancellationToken);
```

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
```
It'd be better to cache stream.CanSeek into a local bool at the beginning of the method rather than making two calls to it.

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
+            if (async)
+            {
+                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
```
We need to pass in CopyToBufferSize explicitly because this lib is targeting .NET Standard 2.0 which doesn't have the overload that doesn't take a buffer size?

bartonjs replied:
I believe that to be the case: https://apisof.net/catalog/System.IO.Stream.CopyToAsync(Stream,CancellationToken)

ericstj replied:
Correct, this overload is not present.
https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=net-5.0#System_IO_Stream_CopyToAsync_System_IO_Stream_System_Threading_CancellationToken_

Do you think this library should copy https://github.com/dotnet/runtime/blob/5c65d891f203618245184fa54397ced0a8ca806c/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L114?  We could factor that into a helper method.

stephentoub replied:
Essentially that's what the logic above that starts the method is trying to do, except that it would then also be used as the argument to CopyTo{Async} in addition to being used to configure the MemoryStream.

ericstj replied:
True, so.
```C#
int bufferSize = canSeek ? Math.Min(streamLength, CopyToBufferSize) : CopyToBufferSize;
```

And in this case I'll provide to CopyTo as well, since it's better than the default on desktop.

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+    /// </summary>
+    public class BinaryData
+    {
+        private const int CopyToBufferSize = 81920;
```
Nit: I think this would be better as a const inside of the method where it's used, and with a comment that it matches the value used implicitly by CopyTo{Async}

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+        /// </returns>
+        public override string ToString()
+        {
+            if (MemoryMarshal.TryGetArray(_bytes, out ArraySegment<byte> data))
+            {
+                return Encoding.UTF8.GetString(data.Array, data.Offset, data.Count);
+            }
+            return Encoding.UTF8.GetString(_bytes.ToArray());
```
This is unnecessarily expensive for memories wrapping things other than arrays.  If this were targeting core, it could just be:
```C#
return Encoding.UTF8.GetString(_bytes.Span);
```
With it targeting .NET Standard 2.0, it can instead be:
```C#
ReadOnlySpan<byte> span = _bytes.Span;
fixed (byte* ptr = span)
{
    return Encoding.UTF8.GetString(ptr, span.Length);
}
```
That could be the entirety of the implementation, or if we wanted to avoid the pinning in the array case, it could just be the fallback, although many encodings are likely to pin the array anyway.

ericstj replied:
It looks like UTF8 always pins here anyway.  Will take this suggestion.

ericstj replied:
Looks like UTF8 is always pinning, I'll update to your suggestion.

stephentoub commented on src/libraries/System.Memory.Data/src/System/IO/ReadOnlyMemoryStream.cs:
```diff
@@ -0,0 +1,168 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
We already have one of these here:
https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/IO/ReadOnlyMemoryStream.cs
Let's reuse the same one and avoid having two.

bartonjs replied:
This copy has ValidateNotClosed calls that the existing one doesn't; so unifying them would require one of

* Removing the "I'm disposed already" exceptions from here
* Adding the "I'm disposed already" exceptions to the existing one (which, IIRC, is never returned to user code and why it skipped them)
* Something with \#if.

Of those, I'd go for adding the ObjectDisposedException to the current copy; but since it's currently only used for some internal networking state (IIRC) there may be conflicting opinions.

stephentoub replied:
The existing one can actually find its way out to user code, but I'm fine with us adding exceptions on disposed.

ericstj replied:
I had to do some work on the common source in order to add the ValidateNotClosed and to make it cross-compile for netstandard2.0.  Stay tuned for the update.

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+                return default;
+            }
+            return data._bytes.Span;
+        }
+        /// <summary>
```
```suggestion
        }
        
        /// <summary>
```

stephentoub commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+        /// <see langword="true" /> if the specified object is equal to the current object; otherwise, <see langword="false" />.
+        /// </returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object? obj)
+        {
+            return ReferenceEquals(this, obj);
+        }
```
```suggestion
        public override bool Equals(object? obj) => ReferenceEquals(this, obj);
```

bartonjs commented on src/libraries/System.Memory.Data/src/System/BinaryData.cs:
```diff
@@ -0,0 +1,262 @@
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
```
Do we use braceless using-dispose statements anywhere other than test code currently?  I'm, personally, not a fan of it because of how it can do (lexical and lifetime) scope creep very easy...

ericstj replied:
Good point.  I prefer braces and will fix this.

GrabYourPitchforks commented on src/libraries/System.Memory.Data/ref/System.Memory.Data.csproj:
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
```
Is the _net461_ TFM needed here? _netstandard2.0_ should be a proper subset.

bartonjs replied:
It lets the net461 ref unify with the GAC through all of the facades, which reduces the xcopy size of an application that uses the package.  All of our OOBs do this now.

ericstj replied:
Everything that we build for `netstandard2.0` we also build for `net461` see https://github.com/dotnet/runtime/issues/1625

ericstj committed:
Respond to feedback

ericstj committed:
Use common ReadOnlyMemoryStream

Make common ReadOnlyMemoryStream support compiling for NETFramework and
NETStandard2.0, and add ObjectDisposedException checking.

ericstj committed:
Add tests cases for null objects/types

ericstj committed:
Fix System.Text.Json reference

ericstj commented:
This should be ready for re-review.  @stephentoub / @bartonjs have a look at what I had to do to get the common ReadOnlyMemoryStream to work.

stephentoub commented on src/libraries/Common/src/System/IO/ReadOnlyMemoryStream.cs:
```diff
@@ -134,5 +172,42 @@ public override Task CopyToAsync(Stream destination, int bufferSize, Cancellatio
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing)
+                {
+                    _isOpen = false;
+                    _content = default;
+                }
+            }
+            finally
+            {
+                base.Dispose(disposing);
+            }
```
You could simplify this to just:
```C#
_isOpen = false;
_content = default;
base.Dispose(disposing);
```

stephentoub commented on src/libraries/Common/src/System/IO/ReadOnlyMemoryStream.cs:
```diff
@@ -96,23 +128,27 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
             return cancellationToken.IsCancellationRequested ?
                 Task.FromCanceled<int>(cancellationToken) :
-                Task.FromResult(Read(new Span<byte>(buffer, offset, count)));
+                Task.FromResult(ReadBuffer(new Span<byte>(buffer, offset, count)));
```
We should probably add a call to EnsureNotClosed before the return statement, so that it still is checked even if cancellation is already requested.

JoshLove-msft replied:
Should this be backported?

ericstj replied:
Both FileSteam and MemoryStream have the current behavior, I did consider this when adding the dispose checks.
https://github.com/dotnet/runtime/blob/630069daca87b7dbfaf91d0bd91a730f5218afe4/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L378-L382
https://github.com/dotnet/runtime/blob/630069daca87b7dbfaf91d0bd91a730f5218afe4/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L386-L387

Do you still think it's better to check disposed before cancelled?

ericstj replied:
@stephentoub please let me know if you still think this should be changed, since the current behavior matches that of our other streams.

stephentoub replied:
> since the current behavior matches that of our other streams

Some of them.  Others it doesn't, e.g.

https://github.com/dotnet/runtime/blob/3c801a8a3b29eabc29ae88e1961c772fda17a1d2/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L369-L378

https://github.com/dotnet/runtime/blob/3c801a8a3b29eabc29ae88e1961c772fda17a1d2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs#L516-L519

I think the right behavior is to flag misuse first.  But as you've highlighted, we're inconsistent, so go ahead with what you have for now.

ericstj replied:
Do you think it makes sense to standardize this behavior?  Does it complicate the code that folks write by throwing here, or should the never be calling when token is cancelled and stream is disposed?  I can imagine some scenario where someone chains a ReadAsync on a continuation and doesn't check CT, but cancels and disposes on another thread.

stephentoub replied:
> should they never be calling when token is cancelled and stream is disposed?

It's fine to call when the token is canceled.  It's not fine to call when the stream is disposed, just like in general it's not fine to use a disposed object.

> Do you think it makes sense to standardize this behavior?

Yes, I think we should subsequently do a pass to check that we're validating stream state before we're checking the cancellation token.

ericstj replied:
I see, since you plan to standardize on that behavior then I'll go ahead and change it now.  Need to reset the CI anyhow.

davidfowl commented on src/libraries/System.Memory.Data/ref/System.Memory.Data.cs:
```diff
@@ -0,0 +1,33 @@
+        public static System.BinaryData FromBytes(System.ReadOnlyMemory<byte> data) { throw null; }
+        public static System.BinaryData FromObjectAsJson<T>(T jsonSerializable, System.Text.Json.JsonSerializerOptions? options = null) { throw null; }
+        public static System.BinaryData FromStream(System.IO.Stream stream) { throw null; }
+        public static System.Threading.Tasks.Task<System.BinaryData> FromStreamAsync(System.IO.Stream stream, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
How come this isn't ValueTask?

GrabYourPitchforks replied:
This was discussed over at https://github.com/Azure/azure-sdk-for-net/pull/16248, but I can't find the comment right now because of a rebase. If I recall, the pattern was that the `ValueTask<T>` APIs were moved to extension methods, allowing this method to be a simple wrapper over the existing Stream methods.

ericstj committed:
Simplify ReadOnlyMemoryStream.Dispose

ericstj committed:
Throw on Disposed for more cases in ReadOnlyMemoryStream
